src/bridge/cockpitfsdir.c

```diff
@@ -214,19 +229,26 @@ cockpit_fsdir_prepare (CockpitChannel *channel)
                                    self->cancellable,
                                    on_enumerator_ready,
                                    self);
-  self->monitor = g_file_monitor_directory (file, 0, NULL, &error);
-  g_object_unref (file);
 
-  if (self->monitor == NULL)
+  if (watch)
     {
-      g_message ("%s: %s", self->path, error->message);
-      options = cockpit_channel_close_options (channel);
-      json_object_set_string_member (options, "message", error->message);
-      problem = "internal-error";
-      goto out;
-    }
+      self->monitor = g_file_monitor_directory (file, 0, NULL, &error);
+      g_object_unref (file);
+      if (self->monitor == NULL)
+        {
+          g_message ("%s: couldn't monitor directory: %s", self->path, error->message);
+          options = cockpit_channel_close_options (channel);
+          json_object_set_string_member (options, "message", error->message);
+          problem = "internal-error";
+          goto out;
+        }
 
-  self->sig_changed = g_signal_connect (self->monitor, "changed", G_CALLBACK (on_changed), self);
+      self->sig_changed = g_signal_connect (self->monitor, "changed", G_CALLBACK (on_changed), self);
+    }
+  else
+    {
+      g_object_unref (file);
```

Comment:
Lets just declare GFile *file = NULL earlier, and run this conditionally in out: 

Hopefully that makes everything simpler.